infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogReplicationSinkManager.java

```diff
@@ -422,8 +425,11 @@ private void processSnapshotMessage(LogReplicationEntry entry) {
         }
     }
 
-    private void startSnapshotApplyAsync(LogReplicationEntry entry) {
-        applyExecutor.submit(() -> startSnapshotApply(entry));
+    private synchronized void startSnapshotApplyAsync(LogReplicationEntry entry) {
+        if (!ongoingApply.get()) {
+            ongoingApply.set(true);
```

Comment:
Is there a race condition here? These two methods are not atomic. Should we use getAndSet()?

Reply:
+1

Reply:
We can't use getAndSet cause we only want to set given the output of the get... But you're right we could have a race condition with the startSnapshotApply so I've synchronized that method as well!